src/vs/workbench/contrib/terminal/browser/xterm/cognisantCommandTrackerAddon.ts

```diff
@@ -151,12 +151,13 @@ export class CognisantCommandTrackerAddon extends CommandTrackerAddon {
 	}
 
 	get cwds(): string[] {
-		const cwds = [];
-		const sorted = new Map([...this._cwds.entries()].sort((a, b) => b[1] - a[1]));
-		for (const [key,] of sorted.entries()) {
-			cwds.push(key);
-		}
-		return cwds;
+		return Array.from(new Map([...this._cwds.entries()].sort((a, b) => a[1] - b[1]))).map(s => s[0]);
+	}
+
+	getCwdForLine(line: number): string {
+		const command = this._commands.sort((a, b) => b.marker!.line - a.marker!.line);
```

Comment:
I guess this is mostly guaranteed right now but not entirely. Ideally we would maintain this._commands as a list sorted by marker.line. To do this I think we just need to check the last entry and if it's not before, splice it into the _commands.

Reply:
I can't think of why the order wouldn't be guaranteed since we store them as they run.

Reply:
going to use unshift instead of push because in both instances when we get commands, the array is reversed

Reply:
The shell could move the cursor above the prompt with this for example:

```
CSI Ps ; Ps H
          Cursor Position [row;column] (default = [1,1]) (CUP).
```

Not expected but it could happen so we should handle it.